### PR TITLE
Add support for connecting to a pool of servers

### PR DESCRIPTION
**Pull request by arnavk**
on `connectionFailed`, tell the pool to try with the next server address. The old messages will get retried on the previous server address till they eventually fail. ¯_(ツ)_/¯

### CODE REVIEW

**Comment by arnavk** [comment]
Add provision to take servers out of rotation for a period of time

File: txconnpool/pool.py [hunk 1]
```diff
@@ -39,12 +38,11 @@ class PooledClientFactory(ReconnectingClientFactory):
     protocol = None
     connectionPool = None
     _protocolInstance = None
-
+    failed_callback = None
```
Comment (arnavk) [inline txconnpool/pool.py]: unused

File: txconnpool/pool.py [hunk 2]
```diff
@@ -163,6 +176,31 @@ def _shutdownCallback(self):
...
+            self._next_server_index = effective_index
+
+            if server_key in self._serverOORRecords:
+                if time.time() > self._serverOORRecords[server_key]:
+                    del self._serverOORRecords[server_key]
+                    return self._serverAddresses[effective_index]
+            else:
+                return self._serverAddresses[effective_index]
+
+        return self._serverAddresses[effective_index]
```
Comment (arnavk) [inline txconnpool/pool.py:197]: log before returning

File: txconnpool/pool.py [hunk 3]
```diff
@@ -183,15 +221,14 @@ def _connected(client):
 
         self._factories.append(factory)
 
-        self._reactor.connectTCP(self._serverAddress.host,
-                                 self._serverAddress.port,
+        self._reactor.connectTCP(self._nextServerAddress.host,
```
Comment (gchiam) [inline txconnpool/pool.py:224]: better to store `self._nextServerAddress` in a local variable and get its host and port, than accessing `self._nextServerAddress` twice to get `host` and `port`

File: txconnpool/pool.py [hunk 4]
```diff
@@ -121,7 +127,14 @@ def __init__(self, serverAddress, maxClients=5,
         @param reactor: An L{IReactorTCP{ provider used to initiate new
             connections.
         """
-        self._serverAddress = serverAddress
+        self._active_server_index = 0
+
+        if not isinstance(serverAddresses, list):
+            serverAddresses = [self.serverAddresses]
```
Comment (arnavk) [inline txconnpool/pool.py:133]: typo. should not be `self.serverAddress`

**Comment by arnavk** [comment]
closing in favour of https://github.com/zopim/txconnpool/pull/1

**Pull request closed without merging**